### ISSUE TITLE
CanJS 5.0

### ISSUE THREAD

**Issue by justinbmeyer** [canjs/canjs#3854]
- [ ] Soft release mostly focused on can-query-logic (site updated) (maybe some minor)
  - [X] Tree shaking CanJS - (5 days)
  - [x] Updating Apps (Bitballs, Bitcentive, etc, updating to use can-query-logic and tree shaking) (2 weeks)
  - [ ] [CanJS Documentation](https://github.com/canjs/canjs/issues/4116) (migration guide, looking for can-set lingering)
  - [x] Tree shaking works with webpack
- [ ] Steal 2.0
- [x] Update "main" docs and `Getting Started` guide for `import {x} from "can"`
- [x] `import {component} from "https://unpkg/can@5.0/all-es.js"` in a JSBin
- [ ] Hard release + Blog article

This is a list of all issues that would require breaking changes.

## Architectural 

These issues are unlikely to break someone's app, but might if someone is relying on low-level behavior that is likely undocumented.

- [ ] [Order observation updates by their depth automatically](https://github.com/canjs/can-observation/issues/125) - Possibly not a breaking change as can-observation is no longer stateful.
- [ ] [Observable lifecycle hooks](https://github.com/canjs/canjs/issues/3855)
- [ ] Remove can-util
- [ ] Use `can-ajax@2.0` places

## User Facing

- [Proposal: Expanded Types and Required Properties](https://github.com/canjs/can-define/issues/287)- Ex: `prop: String`

**Comment by matthewp** [comment]
## Issues to add

* Migration guide
* Advanced setup guide (using packages)

## Immediately after

* All docs use ES imports
* Webpack stache plugin

**Comment by justinbmeyer** [comment]
@matthewp I wrote up some "background" on CanJS's builds.  Here's what I wrote, let me know what you think (cc @phillipskevin / @chasenlehara ):

## Background Information

CanJS has a variety of packages, modules, and files ready to meet the needs of
any development environment. This section gives some background on these items.

- __The individual packages.__

  CanJS is composed from over 80 individual packages.  For example [can-component Component]
  is actually the [can-component](https://www.npmjs.com/package/can-component) package, housed
  in its own [github repository](https://github.com/canjs/can-component). The modules within
  these packages are written in ES5 JavaScript and CommonJS, so they can be imported by Webpack, Browserify,
  StealJS, and a do not require transpiling.

  Apps that need long-term flexibility should installing these packages directly. Direct installation
  means you can upgrade a small part of CanJS when needed.

- __The `can` package.__

  CanJS publishes to [npm](https://www.npmjs.com/) a
  [can package](https://www.npmjs.com/package/can). It contains:

  - `./can.js` - The main export, imports every CanJS sub project and exports it as
    a ES named export. For example, `can.js` looks like this:

    ```js
    export { default as Component } from "can-component";
    export { default as restModel } from "can-rest-model";
    ...
    ```

    Most module loaders with tree-shaking (ex: Webpack and StealJS) setups
    use this module. Import
    named exports from the `can` package like this:

    ```js
    import {Component, restModel} from "can";
    ```

  - `./dist/can.mjs` - An ESM module with named exports of each [core] package bundled
    together. This is useful for examples and prototyping in modern browsers that
    support ES modules and for real-world apps that use just what's in core CanJS.
    It's hosted statically on `unpkg` and can be
    downloaded here.

    You can import directly from the file as follows:

    ```js
    <script type="module">
	import { Component } from "//unpkg.com/can/dist/can.mjs";

    	Component.extend({
    		tag: "my-app",
    		view: `Hello {{name}}!`,
    		ViewModel: {
    			name: { default: "world" }
    		}
    	});

    </script>
    ```

  - `./dist/can.all.min.js` - A minified version of `./dist/can.all.mjs`.

  - `./dist/global/can.js` - A JavaScript file that exports a `can` object with
    core named exports on it. Use this if you want to create a demo or example that will
    work in browsers that do not support ESM.

  - `./dist/can.all.mjs` - An ESM module with named exports of every* package bundled
    together. This is useful for examples and prototyping in modern browsers that
    support ES modules. It's hosted statically on `unpkg` and can be
    downloaded here.

    You can import directly from the file as follows:

    ```js
    <script type="module">
	import { Component } from "//unpkg.com/can/dist/can.all.mjs";

    	Component.extend({
    		tag: "my-app",
    		view: `Hello {{name}}!`,
    		ViewModel: {
    			name: { default: "world" }
    		}
    	});

    </script>
    ```

    This file is huge as it includes nearly every extension to
    CanJS. So use of this module in production, without tree-shaking, is
    not advised!

  - `./dist/can.all.min.js` - A minified version of `./dist/can.all.mjs`.

  - `./dist/global/can.all.js` - A JavaScript file that exports a `can` object with
    every* named export. Use this if you want to create a demo or example that will
    work in browsers that do not support ESM.

=======

Basically, do we want to create a `all` vs `core` build?  I sorta feel like some people will be put off by a 400k download.  I think we should make a paved path for the crowd that just needs "core canjs" for some basic widgets.  I'm even thinking about what would a `can-component`-only build be too.  

Anyhoo .. let me know what you think. Thanks!

**Comment by matthewp** [comment]
I think having a separate core build makes sense. However this does mean we need to have a `core.js` and an `ecosystem.js` like we used to have. Then the `can.js` would just re-export all of those things:

```js
export * from "./core";
export * from "./ecosystem";
export * from "./legacy";
```

I'm lukewarm about adding a minified build. Projects minify their own stuff. It would only be useful if you were using native modules in production.

**Comment by justinbmeyer** [comment]
> I'm lukewarm about adding a minified build. Projects minify their own stuff. It would only be useful if you were using native modules in production.

I think it might be useful for the million hobby and corporate intranet type projects that don't really care about performance / creating a build script. They just want to listen to when a form is submitted and make an ajax request in a slightly nicer way.

**Comment by matthewp** [comment]
We can do it. I don't really buy that argument though, if they don't care about performance then why do they want a minified version.

**Comment by justinbmeyer** [comment]
![image](https://user-images.githubusercontent.com/78602/42849244-0b83608c-89e8-11e8-821f-817b96027d75.png)

**Comment by matthewp** [comment]
It's out 🎆
